File: 2022_11_03_leetcode_in_fsharp/LeetCode/P0387_FirstUniqChar.fs
// url: https://leetcode.com/problems/first-unique-character-in-a-string/
// tags: indexed, tryFindIndex, forall
// examples: "letmein" -> 0

module LeetCode.P0387_FirstUniqChar

open Utils

let firstUniqChar (s: string) =
    s
    |> Seq.indexed
    |> Seq.tryFindIndex (fun (i, c) -> { 0 .. s.Length - 1 } |> Seq.forall (fun j -> i = j || c <> s[j]))
    |> Option.defaultValue -1

firstUniqChar "letmein" === 0
firstUniqChar "lifeislovepoem" === 2
firstUniqChar "aabb" === -1

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0387_FirstUniqChar.cs
// https://leetcode.com/problems/first-unique-character-in-a-string/

namespace LeetCode;

using static System.Linq.Enumerable;

class P0387_FirstUniqChar
{
    static int FirstUniqChar(string s)
    {
        return s
            .Select((c, i) => new { Char = c, Index = i })
            .FirstOrDefault(t => Range(0, s.Length).All(j => t.Index == j || t.Char != s[j]))
            ?.Index ?? -1;
    }
    public static void Run()
    {
        var inputs = new[] { "letmein", "lifeislovepoem", "aabb" };

        foreach (var input in inputs)
        {
            var output = FirstUniqChar(input);
            Console.WriteLine($" {input} -> {output}");
        }
    }
}
